plugins/mysql/mysql_audit

```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/python
+# -*- python -*-
+
+"""
+=head1 INTRODUCTION
+
+Plugin to monitor the MySQL audit log connection count
+
+=head1 INSTALLATION
+
+Place in /etc/munin/plugins/ (or link it there using ln -s)
+
+=head1 CONFIGURATION
+
+Add this to your /etc/munin/plugin-conf.d/munin-node:
+
+=over 2
+
+      [mysql_audit]
+      user mysql
+      period 3
+      logfile /var/log/mysql/audit.log
+      sorted 1 # sort output (if not: unsorted)
+      toplist 10 # only the top 10 (if unset: all of them). implies sorted
+
+=back
+
+NOTE: MySQL needs to be configured manually in order to enable the audit log!
+This can be done as follows:
+
+=over 2
+
+      plugin-load = server_audit=server_audit.so
+      server_audit_events = connect
+      server_audit_output_type=file
+      server_audit_file_path = /var/log/mysql/audit.log
+      server_audit_file_rotate_size = 512M
+      server_audit_logging = ON
+      server_audit_file_rotations = 5
+
+=back
+
+=head1 HISTORY
+
+2017-11-03: v 1.0 Bert Van de Poel <bert@bhack.net>: created
+2020-07-19: v 1.1 pcy <pcy@ulyssis.org>: added config options
+
+=head1 USAGE
+
+Parameters understood:
+
+  config   (required)
+  autoconf (optional - used by munin-config)
+
+=head1 MAGIC MARKERS
+
+#%# family=auto
+#%# capabilities=autoconf
+"""
+
+
+import sys
+import os
+from datetime import datetime, timedelta
+import operator
+
+
+def weakbool(x):
+    return x.lower().strip() in {'true','yes','y','1'}
+
+
+tdminutes = os.getenv('period', '5')
+logfile = os.getenv('logfile', '/var/log/mysql/audit.log')
+toplist = int(os.getenv('toplist', '0'))
+sortlist = weakbool(os.getenv('sorted', 'N')) or toplist > 0
+
+
+def reverse_lines(filename, BUFSIZE=4096):
+    f = open(filename, "r")
+    f.seek(0, 2)
+    p = f.tell()
+    remainder = ""
+    while True:
```

Comment:
After reading the function twice, I understood its purpose, I guess ... :)

I think, a statefile would be suitable here. In order to reduce re-parsing of old lines, the statefile could contain the filesize during the previous run (i.e. the already processed content). Whenever the file size decreases you would start again at the beginning. This would would mimic the implementation of `tail_open` in `Munin::Plugins`.

Reply:
I don't think that extra step is needed here, as the `for` loop further down will break after the date cutoff. All it'd do is reduce the amount of seeking a bit, but not the parsing.

It would be nice to have of course, for other plugins, but I don't feel like messing with this function much... (A colleague of mine wrote it for the shorewall plugin, I just took the code. (Not exactly a colleague as what we're doing isn't technically our job, but, you get the idea.))